geanyvc/src/vc_fossil.c

```diff
@@ -0,0 +1,590 @@
+		got_error = (!*start);
+		if (got_error) break;
+
+		len = end - start;
+		base_name = g_malloc0(len + 1);
+		memcpy(base_name, start, len);
+		filename = g_build_filename(base_dir, base_name, NULL);
+		g_free(base_name);
+
+		item = g_new(CommitItem, 1);
+		item->status = status;
+		item->path = filename;
+
+		lst = g_slist_append(lst, item);
+	}
+
+	if (got_error)
+	{
+		GSList *e;
+		for (e = lst; e ; e = g_slist_next(e))
+		{
+			CommitItem *item = (CommitItem *) (e->data);
+			g_free(item->path);
+			g_free(item);
+		}
+		g_slist_free(lst);
+		lst = NULL;
+	}
+
+	return lst;
+}
+
+static GSList *
+get_commit_files_fossil(const gchar * file)
+{
+	const gchar *argv[] = { FOSSIL_CLIENT, "status", NULL };
+	gchar *std_out = NULL;
+	gchar *base_dir = get_base_dir(file);
+	GSList *ret = NULL;
+
+	g_return_val_if_fail(base_dir, NULL);
+	if (!base_dir) return NULL;
+
+	execute_custom_command(base_dir, (const gchar **) argv, (const gchar **) NULL,
+			       &std_out, NULL, base_dir, NULL, NULL);
+	g_return_val_if_fail(std_out, NULL);
+
+
+	ret = parse_fossil_status(ret, base_dir, std_out, "EDITED", FILE_STATUS_MODIFIED);
+	ret = parse_fossil_status(ret, base_dir, std_out, "UPDATED", FILE_STATUS_MODIFIED);
+	ret = parse_fossil_status(ret, base_dir, std_out, "ADDED", FILE_STATUS_ADDED);
+	ret = parse_fossil_status(ret, base_dir, std_out, "DELETED", FILE_STATUS_DELETED);
+
+	g_free(std_out);
+	g_free(base_dir);
+
+	return ret;
+}
+
+static gint
+fossil_revert_dir(gchar ** std_out, gchar ** std_err, const gchar * path,
+	 GSList * list, const gchar * message)
+{
+	gchar *base_dir = get_base_dir(path);
+    const char **argv = (const gchar **)FOSSIL_CMD_REVERT_DIR;
+	const gchar *base_argv[] = { FOSSIL_CLIENT, "revert", NULL, NULL };
+	gint ret;
+
+	g_return_val_if_fail(base_dir, -1);
+	if (!base_dir) return -1;
+
+	if ( g_strcmp0(path, base_dir) == 0 )
+		argv = (const gchar **)base_argv;
+
+	ret = execute_custom_command(base_dir, argv, (const gchar **) NULL,
+						std_out, std_err, path, list, message);
+	g_free(base_dir);
+	return ret;
+}
+
+static gint
+fossil_status_extra(gchar ** std_out, gchar ** std_err, const gchar * path,
+	 GSList * list, const gchar * message)
+{
+	gchar *base_dir = get_base_dir(path);
+	gint ret = 0;
+
+	g_return_val_if_fail(base_dir, -1);
+	if (!base_dir) return -1;
+
+	ret = execute_custom_command(base_dir, FOSSIL_CMD_STATUS, (const gchar **) NULL,
+						std_out, std_err, path, list, message);
+
+	if (ret)
+	{
+		g_free(base_dir);
+		return ret;
+	}
+
+	if (!ret)
+	{
+		const gchar *argv[] = { FOSSIL_CLIENT, "changes", "--extra", "--classify", NULL };
+		gchar *std_out1 = NULL;
+		gchar *std_err1 = NULL;
+
+		if (std_out)
+		{
+			std_out1 = *std_out;
+			*std_out = NULL;
+		}
+		if (std_err)
+		{
+			std_err1 = *std_err;
+			*std_err = NULL;
+		}
+
+		ret = execute_custom_command(base_dir, (const gchar **) argv, (const gchar **) NULL,
+							std_out, std_err, path, list, message);
+
+		if (std_out1)
```

Comment:
ActuallyI I'm unsure what this is doing.

Reply:
This appends the output of the second command to the output from the first command.

Reply:
Ah. Somehow went confused